### ISSUE TITLE
Crash when using mono 5.4.x

### ISSUE THREAD

**Issue by peter-sabath** [agracio/edge-js#15]
Since the Tomasz seems to be inactive I post that issue here too.

I am trying to get it running with mono 5.4 and encounter a crash (on MacOS).

This crash originates from the Dictionary::Add method, because the 
```
add = mono_class_get_method_from_name(mono_object_get_class(_this), 
"System.Collections.Generic.IDictionary<string,object>.Add", -1);
```

call in line 8 returns a NULL pointer. I understand the reason for this, but don't understand why this does not occur under mono 4.8.

The most common usage is with an `ExpandoObject` which implements `System.Collections.Generic.IDictionary<string,object>.Add` but unfortunately as private interface using explicit interface implementation. Using reflection clearly shows the method to be private.

The methods can be made visible by explicitly casting an object to the interface. While this works when written in C# I am still not able to get it working with mono embedding.

I request the method from the specific interface class <string, object> and use `mono_object_castclass_mbyref`to cast the `_this` object to one of  the interface class but this still does not properly work

My current workaround is an additional static method in MonoEmbedding.cs that gets the object, key & value and makes the operation in C# code, which works, but I am not happy about the switch from native to managed and back which is used many times.

Dictionary.cpp (Changed):
```
void Dictionary::Add(MonoObject* _this, const char* name, MonoObject* value)
{
    static MonoMethod* add;
    
    if(!add) {
        add = mono_class_get_method_from_name(MonoEmbedding::GetClass(), "AddPropertyToObject", -1);
    }

    void* params[3];
    params[0] = _this;	//dict
    params[1] = mono_string_new(mono_domain_get(), name);	// key
    params[2] = value;	// value

    mono_runtime_invoke(add, _this, params, NULL);
}
```

monoembedding.cs (Added new method)
```
    // used for ExpandoObject & IDictionary<string, object> to add values
    static public void AddPropertyToObject( object obj, string propName, object val)
    {
    	var eo = obj as IDictionary<string, object>;
    	if ( eo == null )
    		return;
    		
    	eo.Add(propName, val);
    }
```

I use it in an electron context, but think this is a general problem of both forks.

Anyone with a better solution?

**Comment by agracio** [comment]
You did not specify what Edge.js code you executing while experiencing the crash.
It is also worth pointing out that Mono runtime support has been generally broken since May 5th and May 7th commits into the original repo: https://github.com/tjanczuk/edge/commits/master.

**Comment by peter-sabath** [comment]
The code does not really matter, each operation that involves creating an ExpandoObject (MarshalV8ToCLR) leads to this crash. e.g.

```
const edge = require('electron-edge-js'); 

var helloWorld = edge.func(function () {/*
    async (input) => { 
        return ".NET Welcomes " + input.ToString(); 
    }
*/});
```

This creates an object to pass the to-be-compiled source code.

or any call to a func with a javascript object as parameter.

My code is based on you 8.2.9 version of electron-edge-js.

**Comment by agracio** [comment]
I will look at your proposed changes and see if Mono runtime can be fixed. Will be a few days though.

**Comment by peter-sabath** [comment]
Thanks, with the above outlined approach we have a workaround and most of the team is starting into their Christmas holidays, so there is currently no real hurry.

**Comment by agracio** [comment]
I have merged your changes to the repo and it fixed all test crashes on Mono runtime.
However I am facing problems with Mono compiling on the latest Node.Js 9.3.0 and was wondering if you could give me a hand.

Errors:
```.0.
../src/mono/clrfuncinvokecontext.cpp: In member function ‘v8::Local<v8::Value> ClrFuncInvokeContext::CompleteOnV8Thread(bool)’:
../src/mono/clrfuncinvokecontext.cpp:103:13: error: ‘Task’ is not a class, namespace, or enumeration
     switch (Task::Status(this->Task())) {
             ^
../src/mono/clrfuncinvokecontext.cpp:108:13: error: ‘Task’ is not a class, namespace, or enumeration
         if (Task::Exception(this->Task()) != NULL) {
             ^
../src/mono/clrfuncinvokecontext.cpp:109:56: error: ‘Task’ is not a class, namespace, or enumeration
             argv[0] = ClrFunc::MarshalCLRExceptionToV8(Task::Exception(this->Task()));
                                                        ^
../src/mono/clrfuncinvokecontext.cpp:121:43: error: ‘Task’ is not a class, namespace, or enumeration
         argv[1] = ClrFunc::MarshalCLRToV8(Task::Result(this->Task()), &exc);
```
Full error stack: https://travis-ci.org/agracio/edge-js/jobs/318308987

Some background to this failure - it looks like new V8 definition introduced its own class `v8::Task`, this also caused problems on .NET Framework compile where i was getting errors like 
```
..\src\dotnet\clrfunc.cpp(518): error C2872: 'Task': ambiguous symbol [C:\Users\anton\Dev\edge-js\build\edge_nativeclr.
vcxproj]
  ..\src\dotnet\clrfunc.cpp(518): note: could be 'System::Threading::Tasks::Task'
  c:\users\anton\.node-gyp\9.3.0\include\node\v8-platform.h(20): note: or       'v8::Task' (compiling source file ..\sr
  c\dotnet\clrfunc.cpp)
```
This was easy to fix by using a full qualifier `System::Threading::Tasks::Task` instead of `Task` https://github.com/agracio/edge-js/commit/e7bd5f9b9665cd8717ec64bdd3a98b6114c4765a#diff-c52d7b9918e39cc1a6572a363ebea2e6L35.

But I cannot get my head around the Mono problem and my C++ skills are lacking to put it mildly.
So if you have some spare time could you take a look at it. 
Node.js versions 6.x.x, 7.x.x and 8.x.x are not affected by this problem and can be safely used with Mono and version 9.0.0 was also compiling fine.

The new version is not in NPM yet but will publish in a few days, just hoping to get this issue out of the way.

**Comment by agracio** [comment]
Fixed all issues on Mono now, new releases that includes your code changes and fixes to Mono runtime on 9.3.0 is now on NPM
